packages/fuels-core/src/types/transaction_builders.rs

```diff
@@ -252,48 +288,146 @@ impl ScriptTransactionBuilder {
+            self.base_offset()
+        } else {
+            0
+        };
+
+        let num_witnesses = self.num_witnesses()?;
+        let tx = self
+            .resolve_fuel_tx_provider(base_offset, num_witnesses, provider)
+            .await?;
+
+        Ok(ScriptTransaction {
+            tx,
+            is_using_predicates,
+        })
+    }
+
+    // When dry running a tx with `utxo_validation` off, the node will not validate signatures.
+    // However, the node will check if the right number of witnesses is present.
+    // This function will create empty witnesses such that the total length matches the expected one.
+    fn create_dry_run_witnesses(&self, num_witnesses: u8) -> Vec<Witness> {
+        let unresolved_witnesses_len = self.unresolved_signatures.addr_idx_offset_map.len();
+        repeat(Default::default())
+            // Add one in case there is no witnesses at all
+            .take(num_witnesses as usize + unresolved_witnesses_len + 1)
+            .collect()
+    }
+
+    // When the `script_gas_limit` was not set by the user, `dry_run` the tx
+    // and set the `script_gas_limit` to the actual `gas_used`
+    async fn set_script_gas_limit_to_gas_used(
+        tx: &mut Script,
+        provider: &impl DryRunner,
+        network_info: &NetworkInfo,
+        tolerance: f32,
+    ) -> Result<()> {
+        let consensus_params = &network_info.consensus_parameters;
+        // Add `1` because of rounding
+        let max_gas = tx.max_gas(consensus_params.gas_costs(), consensus_params.fee_params()) + 1;
+
+        // If `gas_limit` was not set use `max_gas_per_tx` but subtract the tx's base maximum cost
+        // TODO: @xgreenx why do I need to / 2
+        tx.set_script_gas_limit((network_info.max_gas_per_tx / 2) - max_gas);
+
+        // The `dry_run` validation will check if there is an input present that can cover
+        // the tx fees. If we are estimating without inputs we have to add a temporary one
+        tx.inputs_mut().push(FuelInput::coin_signed(
+            Default::default(),
+            Default::default(),
+            1_000_000_000,
+            Default::default(),
+            TxPointer::default(),
+            0,
+            0u32.into(),
+        ));
+
+        let gas_used = provider
+            .dry_run_and_get_used_gas(tx.clone().into(), tolerance)
+            .await?;
 
+        // Remove the temporary coin
+        tx.inputs_mut().pop();
+
+        tx.set_script_gas_limit(gas_used);
+
+        Ok(())
+    }
+
+    async fn resolve_fuel_tx_provider(
+        self,
+        base_offset: usize,
+        num_witnesses: u8,
+        provider: &impl DryRunner,
+    ) -> Result<Script> {
+        let mut policies = self.generate_shared_fuel_policies();
+        policies.set(
+            PolicyType::WitnessLimit,
+            self.witness_limit.or(Some(DEFAULT_SCRIPT_WITNESS_LIMIT)),
+        );
+
+        let script_len = self.script.len();
```

Comment:
Maybe `let no_code_present = self.script.is_empty()`

would eliminate the need for the comment below.